scheds/rust/scx_cosmos/src/bpf/main.bpf.c

```diff
@@ -234,6 +239,69 @@ static inline u64 shared_dsq(s32 cpu)
 	return numa_enabled ? __COMPAT_scx_bpf_cpu_node(cpu) : SHARED_DSQ;
 }
 
+/*
+ * Return true if @p can only run on a single CPU, false otherwise.
+ */
+static inline bool is_pcpu_task(const struct task_struct *p)
+{
+	return p->nr_cpus_allowed == 1 || is_migration_disabled(p);
+}
+
+/*
+ * Return true if the system is busy, false otherwise.
+ *
+ * This function determines when the scheduler needs to switch to
+ * deadline-mode (using a shared DSQ) vs round-robin mode (using per-CPU
+ * local DSQs).
+ */
+static inline bool is_system_busy(void)
+{
+	return cpu_util >= busy_threshold;
+}
+
+/*
+ * Return true if the CPU is running the idle thread, false otherwise.
+ */
+static inline bool is_cpu_idle(s32 cpu)
+{
+	return scx_bpf_cpu_rq(cpu)->curr->flags & PF_IDLE;
+}
+
+/*
+ * Find an idle CPU iterating over the available CPUs, instead of relying
+ * on the idle cpumasks.
+ *
+ * This method might be inaccurate, but it's less expensive than operating
+ * on cpumasks with simple topologies, especially under light or moderate
+ * load.
+ */
+static s32 pick_idle_cpu_flat(struct task_struct *p, s32 prev_cpu)
+{
+	static s32 last_cpu;
+	s32 i;
+
+	if (is_cpu_idle(prev_cpu))
+		return prev_cpu;
+
+	if (is_pcpu_task(p))
+		return -EBUSY;
+
+	bpf_for(i, 0, nr_cpu_ids) {
+		s32 cpu = (last_cpu + i) % nr_cpu_ids;
```

Comment:
Is it intended to use the moving `last_cpu` here? It seems like the loop intends to cover every CPU, but that won't happen if this method races. It seems redundant to still loop over all the CPUs even though other racers have already checked they're idle.